### PR TITLE
Refactoring

### PR DESCRIPTION
**Pull request by isqua**

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -24,6 +24,8 @@ module.exports = postcss.plugin(
   function fixUrl(options) {
     options = options || {}
     var mode = options.url !== undefined ? options.url : "rebase"
+    var isCustom = typeof mode === "function"
+    var callback = isCustom ? getCustomProcessor(mode) : getUrlProcessor(mode)
```
Comment (isqua) [inline index.js]: First of all, get `url-processor-function` one time. Earlier it was gotten for each declaration.

File: index.js [hunk 2]
```diff
@@ -33,9 +35,11 @@ module.exports = postcss.plugin(
         ? path.dirname(result.opts.to)
         : from
 
+      var cb = getDeclProcessor(result, from, to, callback, options, isCustom)
```
Comment (isqua) [inline index.js]: Next, I create a closure function to save parameters, which are equal for all declarations.

File: index.js [hunk 3]
```diff
@@ -33,9 +35,11 @@ module.exports = postcss.plugin(
...
+
       styles.walkDecls(function(decl) {
         if (decl.value && decl.value.indexOf("url(") > -1) {
-          processDecl(result, decl, from, to, mode, options)
+          processDecl(decl, cb)
```
Comment (isqua) [inline index.js]: ... for passing only decl and callback here

File: index.js [hunk 4]
```diff
@@ -93,69 +125,71 @@ function createUrl(urlMeta, newPath) {
...
-      return processRebase(result, from, dirname, urlMeta, to)
-    case "inline":
-      return processInline(result, from, dirname, urlMeta, to, options, decl)
-    case "copy":
-      return processCopy(result, from, dirname, urlMeta, to, options, decl)
-    default:
-      throw new Error("Unknow mode for postcss-url: " + mode)
-    }
-  })
+    return callback(result, from, dirname, urlMeta, to, options, decl)
```
Comment (isqua) [inline index.js]: All url processors have the same API, even custom callback!

File: index.js [hunk 5]
```diff
@@ -93,69 +125,71 @@ function createUrl(urlMeta, newPath) {
...
+ * @return {PostcssUrl~UrlProcessor}
  */
-function processCustom(result, cb, from, dirname, urlMeta, to, options, decl) {
-  var newValue = cb(urlMeta.value, decl, from, dirname, to, options, result)
-  return createUrl(urlMeta, newValue)
+function getCustomProcessor(cb) {
+  return function(result, from, dirname, urlMeta, to, options, decl) {
+    var newValue = cb(urlMeta.value, decl, from, dirname, to, options, result)
+    return createUrl(urlMeta, newValue)
+  }
```
Comment (isqua) [inline index.js]: Because I wrap it into closure, to make API like in other functions.

**Comment by isqua** [comment]
@MoOx I can split this PR in some smaller PRs, if you want.

**Comment by MoOx** [comment]
Good job !
If all tests pass, I understand that there is no breaking changes or improvement? So no release for now?

**Comment by isqua** [comment]
There is no breaking changes, but it may work faster than before on big files. Take a moment, I estimate tests passing time now and before.

**Comment by isqua** [comment]
Tests is rather small, so its time is about 0,33...0,4s in both branches. So no release I think.

**Comment by MoOx** [comment]
Cool. Thanks!

**Pull request merged**